EduNLP/Pretrain/elmo_vec.py

```diff
@@ -0,0 +1,213 @@
+    ... 若$x,y$满足约束条件公式$\\FormFigureBase64{wrong2?}$,$\\SIFSep$，则$z=x+7 y$的最大值为$\\SIFBlank$"]
+    >>> vocab.tokenize(items[0])
+    ['公式', '如图', '[FIGURE]', 'x', ',', 'y', '约束条件', '公式', '[SEP]', 'z', '=', 'x', '+', '7', 'y', '最大值', '[MARK]']
+    >>> len(vocab)
+    18
+    """
+
+    def __init__(self, path=None):
+        self.pure_tokenizer = PureTextTokenizer()
+        self.t2id = {PAD_SYMBOL: 0, UNK_SYMBOL: 1, FORMULA_SYMBOL: 2, FIGURE_SYMBOL: 3,
+                     QUES_MARK_SYMBOL: 4, TAG_SYMBOL: 5, SEP_SYMBOL: 6}
+        if path is None:
+            pass
+        else:
+            self.load_vocab(path)
+
+    def __call__(self, item, *args, **kwargs):
+        return self.to_index(self.tokenize(item))
+
+    def __len__(self):
+        return len(self.t2id)
+
+    def tokenize(self, item: str):
+        tokens = next(self.pure_tokenizer([item]))
+        for token in tokens:
+            self.append(token)
+        return tokens
+
+    def to_index(self, item: list, max_length=128, pad_to_max_length=False):
+        ret = [self.t2id[UNK_SYMBOL] if token not in self.t2id else self.t2id[token] for token in item]
+        if pad_to_max_length:
+            if len(ret) < max_length:
+                ret = ret + (max_length - len(ret)) * [self.t2id[PAD_SYMBOL]]
+            else:
+                ret = ret[0:max_length - 1]
+        return ret
+
+    def append(self, item):
+        if item in self.t2id:
+            pass
+        else:
+            self.t2id[item] = len(self.t2id)
+
+    def save_vocab(self, path):
+        with open(path, 'w') as f:
+            json.dump(self.t2id, f)
+        return path
+
+    def load_vocab(self, path):
+        with open(path, 'r') as f:
+            self.t2id = json.load(f)
+        return path
+
+
+class ElmoDataset(tud.Dataset):
+    def __init__(self, texts: list, tokenizer: ElmoTokenizer, max_length=128):
+        super(ElmoDataset, self).__init__()
+        self.tokenizer = tokenizer
+        self.texts = [text if len(text) < max_length else text[0:max_length - 1] for text in texts]
+
+    def __len__(self):
+        return len(self.texts)
+
+    def __getitem__(self, index):
+        text = self.texts[index]
+        sample = {
+            'length': len(text),
+            'idx': self.tokenizer.to_index(text, pad_to_max_length=True)
+        }
+        return sample
+
+
+def elmo_collate_fn(batch_data):
+    # batch_data = [torch.tensor(t).cuda() for t in batch_data]
+    # batch_data = torch.nn.utils.rnn.pad_sequence(batch_data)
+    pred_mask = []
+    idx_mask = []
+    max_len = max([data['length'] for data in batch_data])
+    for data in batch_data:
+        pred_mask.append([True] * data['length'] + [False] * (max_len - data['length']))
+    for data in batch_data:
+        idx_mask.append([True] * data['length'] + [False] * (len(data['idx']) - data['length']))
+    ret_batch = {
+        'pred_mask': torch.tensor(pred_mask),
```

Comment:
Actually they are different, because `max_len` is the maximum of real sequence length, while `len(data['idx'])` is the padded sequence length.
Here is an example:
```
sample1 = {'length':4, 'idx':[1,2,3,4,0,0]}
sample2 = {'length':3, 'idx':[7,8,9,0,0,0]}
batch_data = [sampel1, sample2]
```
In elmo_collate_fn:
```
max_len = max(4, 3) = 4
pred_mask = [
[true, true, true, true], 
[true, true, true, false]
]
idx_mask = [
[true, true, true, true, false, false],
[true, true, true, false, false, false]
]
```